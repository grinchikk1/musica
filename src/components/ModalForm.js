import React from "react";
import { Formik, Form, Field, ErrorMessage } from "formik";
import { PatternFormat } from "react-number-format";
import * as Yup from "yup";
import Button from "./Button";
import "../styles/ModalForm.scss";

const validationSchema = Yup.object({
  firstName: Yup.string().required("Введіть ім'я"),
  lastName: Yup.string().required("Введіть прізвище"),
  age: Yup.number().required("Введіть вік").positive("Введіть додатнє число"),
  address: Yup.string().required("Введіть адресу"),
  phone: Yup.string()
    .required("Введіть номер телефону")
    .matches(
      /^\(\d{3}\)\d{3}-\d{2}-\d{2}$/,
      "Невірний номер телефону. Введіть у форматі (###)###-##-##"
    )
    .test("complete-phone", "Введіть повний номер", (value) => {
      return value && value.replace(/[^0-9]/g, "").length === 10;
    }),
});

export default function ModalForm({
  cart,
  closeButton,
  onRequestClose,
  isOpen,
  handleModalClose,
  deleteAllFromCart,
}) {
  if (!isOpen) {
    return null;
  }

  const onSubmit = async (values) => {
    deleteAllFromCart();
    handleModalClose();
    // console.log("Processing order...");
    // await new Promise((resolve) => setTimeout(resolve, 1500));
    console.log("Order completed!");
    console.group("Придбані товари:");
    console.log(cart.map((cart) => cart.title).join(", "));
    console.groupEnd();
    console.group("Інформація про користувача:");
    console.log(
      [
        values.lastName,
        values.firstName,
        values.age,
        values.address,
        values.phone,
      ].join(", ")
    );
    console.groupEnd();
  };

  return (
    <div className="modal-overlay" onClick={onRequestClose}>
      <div className="modal" onClick={(e) => e.stopPropagation()}>
        <div className="modal-header">
          <h2>Заповніть форму</h2>
          {closeButton && (
            <button className="modal-close" onClick={onRequestClose}>
              &times;
            </button>
          )}
        </div>
        <Formik
          initialValues={{
            firstName: "",
            lastName: "",
            age: "",
            address: "",
            phone: "",
          }}
          validationSchema={validationSchema}
          onSubmit={onSubmit}
          validateOnMount
        >
          <Form className="modal-form-content">
            <div className="modal-form-box-field">
              <div className="modal-form-item">
                <label htmlFor="firstName">Ім'я: </label>
                <Field type="text" id="firstName" name="firstName" />
                <ErrorMessage
                  name="firstName"
                  component="div"
                  style={{ color: "red" }}
                />
              </div>
              <div className="modal-form-item">
                <label htmlFor="lastName">Прізвище: </label>
                <Field type="text" id="lastName" name="lastName" />
                <ErrorMessage
                  name="lastName"
                  component="div"
                  style={{ color: "red" }}
                />
              </div>
              <div className="modal-form-item">
                <label htmlFor="age">Вік: </label>
                <Field type="number" id="age" name="age" />
                <ErrorMessage
                  name="age"
                  component="div"
                  style={{ color: "red" }}
                />
              </div>
              <div className="modal-form-item">
                <label htmlFor="address">Адреса: </label>
                <Field type="text" id="address" name="address" />
                <ErrorMessage
                  name="address"
                  component="div"
                  style={{ color: "red" }}
                />
              </div>
              <div className="modal-form-item">
                <label htmlFor="phone">Номер телефону: </label>
                <Field type="text" id="phone" name="phone">
                  {({ form }) => {
                    return (
                      <PatternFormat
                        format="(###)###-##-##"
                        mask="#"
                        id="phone"
                        name="phone"
                        type="tel"
                        placeholder="(###)###-##-##"
                        onValueChange={(values) => {
                          form.setFieldValue("phone", values.formattedValue);
                        }}
                      />
                    );
                  }}
                </Field>
                <ErrorMessage
                  name="phone"
                  component="div"
                  style={{ color: "red" }}
                />
              </div>
            </div>
            <Button
              type="submit"
              backgroundColor="goldenrod"
              text="Checkout"
              onClick={() => {}}
            />
          </Form>
        </Formik>
      </div>
    </div>
  );
}
