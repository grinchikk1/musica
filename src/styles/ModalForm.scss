.modal-form-content {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  line-height: 26px;
  font-size: 14px;
  padding: 30px;
}

.modal-form-box-field {
  display: flex;
  flex-direction: column;
  padding-bottom: 15px;
  gap: 7px;
}

.modal-form-item {
  display: flex;
  flex-direction: column;
}
