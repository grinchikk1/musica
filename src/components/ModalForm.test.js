import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, screen } from "@testing-library/react";
import ModalForm from "./ModalForm";

describe("ModalForm", () => {
  // snapshot test
  it("renders a form with correct fields", () => {
    const tree = renderer.create(<ModalForm isOpen={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("ModalForm validation", () => {
    const cart = [
      { id: 1, name: "Item 1" },
      { id: 2, name: "Item 2" },
    ];
    render(
      <ModalForm
        cart={cart}
        closeButton={true}
        onRequestClose={() => {}}
        handleModalClose={() => {}}
        isOpen={true}
        deleteAllFromCart={() => {}}
      />
    );

    expect(screen.getByText("Ім'я:")).toBeInTheDocument();
    expect(screen.getByText("Прізвище:")).toBeInTheDocument();
    expect(screen.getByText("Вік:")).toBeInTheDocument();
    expect(screen.getByText("Адреса:")).toBeInTheDocument();
    expect(screen.getByText("Номер телефону:")).toBeInTheDocument();

    fireEvent.change(screen.getByLabelText("Ім'я:"), {
      target: { value: "Dmytro" },
    });
    fireEvent.change(screen.getByLabelText("Прізвище:"), {
      target: { value: "Shyk" },
    });
    fireEvent.change(screen.getByLabelText("Вік:"), {
      target: { value: 30 },
    });
    fireEvent.change(screen.getByLabelText("Адреса:"), {
      target: { value: "Main Street" },
    });
    fireEvent.change(screen.getByLabelText("Номер телефону:"), {
      target: { value: "(123)456-78-90" },
    });

    fireEvent.click(screen.getByText("Checkout"));
  });
});
