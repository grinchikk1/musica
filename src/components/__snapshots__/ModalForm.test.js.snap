// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalForm renders a form with correct fields 1`] = `
<div
  className="modal-overlay"
>
  <div
    className="modal"
    onClick={[Function]}
  >
    <div
      className="modal-header"
    >
      <h2>
        Заповніть форму
      </h2>
    </div>
    <form
      action="#"
      className="modal-form-content"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="modal-form-box-field"
      >
        <div
          className="modal-form-item"
        >
          <label
            htmlFor="firstName"
          >
            Ім'я: 
          </label>
          <input
            id="firstName"
            name="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="modal-form-item"
        >
          <label
            htmlFor="lastName"
          >
            Прізвище: 
          </label>
          <input
            id="lastName"
            name="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="modal-form-item"
        >
          <label
            htmlFor="age"
          >
            Вік: 
          </label>
          <input
            id="age"
            name="age"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value=""
          />
        </div>
        <div
          className="modal-form-item"
        >
          <label
            htmlFor="address"
          >
            Адреса: 
          </label>
          <input
            id="address"
            name="address"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="modal-form-item"
        >
          <label
            htmlFor="phone"
          >
            Номер телефону: 
          </label>
          <input
            id="phone"
            inputMode="numeric"
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseUp={[Function]}
            placeholder="(###)###-##-##"
            type="tel"
            value=""
          />
        </div>
      </div>
      <button
        className="button"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "goldenrod",
          }
        }
      >
        Checkout
      </button>
    </form>
  </div>
</div>
`;
